pkg/executor/internal/mpp/local_mpp_coordinator.go

```diff
@@ -343,6 +351,130 @@ func (c *localMppCoordinator) fixTaskForCTEStorageAndReader(exec *tipb.Executor,
+	tikvStore, ok := store.(helper.Storage)
+	if !ok {
+		return
+	}
+	h.tidbZone = config.GetGlobalConfig().Labels[placement.DCLabelKey]
+	cache := tikvStore.GetRegionCache()
+	allTiFlashStores := cache.GetTiFlashStores(tikv.LabelFilterNoTiFlashWriteNode)
+	h.allTiflashZoneInfo = make(map[string]string, len(allTiFlashStores))
+	for _, tiflashStore := range allTiFlashStores {
+		tiflashStoreAddr := tiflashStore.GetAddr()
+		if tiflashZone, isSet := tiflashStore.GetLabelValue(placement.DCLabelKey); isSet {
+			h.allTiflashZoneInfo[tiflashStoreAddr] = tiflashZone
+		}
+	}
+	// initial capacity to 2, for one exchange sender and one exchange receiver
+	h.exchangeZoneInfo = make(map[string][]string, 2)
+}
+
+func (h *taskZoneInfoHelper) tryQuickFillWithUncertainZones(exec *tipb.Executor, slots int, sameZoneFlags []bool) (bool, []bool) {
+	if exec.ExecutorId == nil || len(h.currentTaskZone) == 0 {
+		for i := 0; i < slots; i++ {
+			sameZoneFlags = append(sameZoneFlags, true)
+		}
+		return true, sameZoneFlags
+	}
+	if h.isRoot && exec.Tp == tipb.ExecType_TypeExchangeSender {
+		sameZoneFlags = append(sameZoneFlags, len(h.tidbZone) == 0 || h.currentTaskZone == h.tidbZone)
+		return true, sameZoneFlags
+	}
+	return false, sameZoneFlags
+}
+
+func (h *taskZoneInfoHelper) collectExchangeZoneInfos(encodedTaskMeta [][]byte, slots int) []string {
+	zoneInfos := make([]string, 0, slots)
+	for _, taskBytes := range encodedTaskMeta {
+		taskMeta := &mpp.TaskMeta{}
+		err := taskMeta.Unmarshal(taskBytes)
+		if err != nil {
+			zoneInfos = append(zoneInfos, "")
+			continue
+		}
+		zoneInfos = append(zoneInfos, h.allTiflashZoneInfo[taskMeta.GetAddress()])
+	}
+	return zoneInfos
+}
+
+func (h *taskZoneInfoHelper) fillSameZoneFlagForExchange(exec *tipb.Executor) {
+	children := make([]*tipb.Executor, 0, 2)
+	switch exec.Tp {
+	case tipb.ExecType_TypeTableScan, tipb.ExecType_TypePartitionTableScan, tipb.ExecType_TypeIndexScan:
+	case tipb.ExecType_TypeSelection:
+		children = append(children, exec.Selection.Child)
+	case tipb.ExecType_TypeAggregation, tipb.ExecType_TypeStreamAgg:
+		children = append(children, exec.Aggregation.Child)
+	case tipb.ExecType_TypeTopN:
+		children = append(children, exec.TopN.Child)
+	case tipb.ExecType_TypeLimit:
+		children = append(children, exec.Limit.Child)
+	case tipb.ExecType_TypeExchangeSender:
+		children = append(children, exec.ExchangeSender.Child)
+		slots := len(exec.ExchangeSender.EncodedTaskMeta)
+		sameZoneFlags := make([]bool, 0, slots)
+		filled := false
+		if filled, sameZoneFlags = h.tryQuickFillWithUncertainZones(exec, slots, sameZoneFlags); filled {
+			break
+		}
+		zoneInfos, exist := h.exchangeZoneInfo[*exec.ExecutorId]
+		if !exist {
+			zoneInfos = h.collectExchangeZoneInfos(exec.ExchangeSender.EncodedTaskMeta, slots)
+			h.exchangeZoneInfo[*exec.ExecutorId] = zoneInfos
+		}
+		for i := 0; i < slots; i++ {
+			sameZoneFlags = append(sameZoneFlags, len(zoneInfos[i]) == 0 || h.currentTaskZone == zoneInfos[i])
```

Comment:
When will `len(zoneInfos[i])` be 0?

Reply:
when the respond tiflash node doesn't have zone label, in this situation, "" will be filled.